quantum/painter/qp.c

```diff
@@ -126,44 +126,124 @@ bool qp_flush(painter_device_t device) {
+        return 0;
+    }
+
+    uint16_t width;
     switch (driver->rotation) {
         default:
         case QP_ROTATION_0:
         case QP_ROTATION_180:
-            if (width) {
-                *width = driver->panel_width;
-            }
-            if (height) {
-                *height = driver->panel_height;
-            }
-            break;
+            width = driver->panel_width;
+
         case QP_ROTATION_90:
         case QP_ROTATION_270:
-            if (width) {
-                *width = driver->panel_height;
-            }
-            if (height) {
-                *height = driver->panel_width;
-            }
-            break;
+            width = driver->panel_height;
+    }
+
+    qp_dprintf("qp_get_width: ok\n");
+    return width;
+}
+
+uint16_t qp_get_height(painter_device_t device) {
+    qp_dprintf("qp_get_height: entry\n");
+    painter_driver_t *driver = (painter_driver_t *)device;
+
+    if (!driver || !driver->validate_ok) {
+        qp_dprintf("qp_get_height: fail (validation_ok == false)\n");
+        return 0;
+    }
+
+    uint16_t height;
+    switch (driver->rotation) {
+        default:
+        case QP_ROTATION_0:
+        case QP_ROTATION_180:
+            height = driver->panel_height;
+
+        case QP_ROTATION_90:
+        case QP_ROTATION_270:
+            height = driver->panel_width;
+    }
+
+    qp_dprintf("qp_get_height: ok\n");
+    return height;
+}
+
+painter_rotation_t qp_get_rotation(painter_device_t device) {
+    qp_dprintf("qp_get_rotation: entry\n");
+    painter_driver_t *driver = (painter_driver_t *)device;
+
+    if (!driver || !driver->validate_ok) {
+        qp_dprintf("qp_get_rotation: fail (validation_ok == false)\n");
+        return QP_ROTATION_0;
+    }
+
+    painter_rotation_t rotation = driver->rotation;
+
+    qp_dprintf("qp_get_rotation: ok\n");
+    return rotation;
+}
+
+uint16_t qp_get_offset_x(painter_device_t device) {
+    qp_dprintf("qp_get_offset_x: entry\n");
+    painter_driver_t *driver = (painter_driver_t *)device;
+
+    if (!driver || !driver->validate_ok) {
+        qp_dprintf("qp_get_offset_x: fail (validation_ok == false)\n");
+        return 0;
+    }
+
+    painter_rotation_t offset_x = driver->offset_x;
+
+    qp_dprintf("qp_get_offset_x: ok\n");
+    return offset_x;
+}
+
+uint16_t qp_get_offset_y(painter_device_t device) {
+    qp_dprintf("qp_get_offset_y: entry\n");
+    painter_driver_t *driver = (painter_driver_t *)device;
+
+    painter_rotation_t offset_y = driver->offset_y;
+
+    if (!driver || !driver->validate_ok) {
+        qp_dprintf("qp_get_offset_y: fail (validation_ok == false)\n");
+        return 0;
+    }
+
+    qp_dprintf("qp_get_offset_y: ok\n");
+    return offset_y;
+}
+
+void qp_get_geometry(painter_device_t device, uint16_t *width, uint16_t *height, painter_rotation_t *rotation, uint16_t *offset_x, uint16_t *offset_y) {
+    qp_dprintf("qp_geometry: entry\n");
+
```

Comment:
Add guard clause here too?

Disable debug temporarily to only show this function (and not the little ones) output?

Reply:
No need for a driver validation guard clause here, nor in any of the other functions you've added.

The guard clauses check to make sure that functions actually exist before allowing calls to them. The generic qp drivers have placeholders for certain functions, and when you create a real device/driver (like with `qp_st7735_make_spi_device`) it then assigns real functions (taken from the underlying driver, containing device-specific implementation code) to those placeholder functions.

A good example of this distinction (of when guard code is needed vs not needed) can be seen by comparing `qp_viewport` and `qp_set_viewport_offsets` within `qp.c`. Both functions are part of the external api, but driver validation checks are only performed for `qp_viewport`. The reason is that `qp_set_viewport_offsets` doesn't need to call any of the underlying driver code (that is, any code specific to the particular device you are using) in its implementation, it can just check values like `driver->offset_x` (which can be looked up without any care to what the actual underlyingdevice is, it could be an OLED or a TFT or a surface). In contrast, calling `qp_viewport` has an implementation that changes depending on which device is used, which you can see by how that function calls `driver->driver_vtable->viewport(device, left, top, right, bottom);` (where `viewport` is a function implemented by the specific device driver)

If there is any possibility that a function in the generic driver does not have an implementation tied to it, then it should be checked with the guard function (i.e. validated) to ensure the function isn't null (that is, unassigned). Put another way: the validation functions at the start of `qp.c` check for existence of certain functions; if your implementation calls any of the functions in those lists, it should validate the functions first (otherwise you'd risk calling a function that simply doesn't exist).

As for this particular PR, none of the functions you added call any of the functions in the list checked by the validator, so there is no need to perform validation checks for any of the functions you added. These functions will always exist and be valid, even if no driver is assigned. So you can remove the validation checks (e.g. lines 160-163 in `qp.c`)

Reply:
> No need for a driver validation guard clause here, nor in any of the other functions you've added.
> The guard clauses check to make sure that functions actually exist ....

The `validate_ok` check is indeed not needed but it doesn't harm either, and would confirm that the driver has been inited (these values have been set) before using the getters.
- It's set to false when created
- true when init'ed

 And the `!driver` check is definitely needed for `painter_device_t`'s that haven't been assigned with a `make_*` function yet, otherwise code will crash trying to dereference `NULL`

Reply:
Ahh yes, you are correct. The `!driver` check is indeed necessary (and should probably be added to functions like `qp_set_viewport_offsets` as well). And it is true that checking `validate_ok` is harmless--while there is no risk of dereferencing a `NULL` if validation fails, failing validation still shouldn't happen...and if it does fail for some reason, that's a good indication that something has gone horribly wrong somewhere along the way.


> Add guard clause here too?

Yes, I think if the function is exposed as external API, it should have dedicated guard code and not rely on the guard code of functions it calls, so that it is clear which function call actually triggered the failure.

> Disable debug temporarily to only show this function (and not the little ones) output?

This should not be necessary; if the guard code for `qp_get_geometry` fails, it should exit the function immediately, so the other functions aren't going to be called anyway

Reply:
#20481 for what is worth :P